cpp/bench/ann/src/common/benchmark.hpp

```diff
@@ -287,15 +287,14 @@ void bench_search(::benchmark::State& state,
     std::make_shared<buf<std::size_t>>(current_algo_props->query_memory_type, k * query_set_size);
 
   cuda_timer gpu_timer;
-  auto start = std::chrono::high_resolution_clock::now();
   {
     nvtx_case nvtx{state.name()};
+    [[maybe_unused]] auto ntx_lap = nvtx.lap();
+    [[maybe_unused]] auto gpu_lap = gpu_timer.lap();
```

Comment:
Please don't move the `.lap()` functions out of the loop: as the name suggests, they should be invoked each iteration.
1. NVTX: this marks the benchmark iterations with the NVTX ranges in the nsys timeline; moving this out of the loop breaks the marks. If you really like to avoid this (minimal) overhead, just disable the NVTX reporting when calling the benchmark with nsys.
2. GPU: this serves not only as the timer but also as a synchronization point. Without that, the GPU work in the `gpu_timer.stream()` spills out of the iteration section. That is the lack of synchronization what may give you the observed speedup, not the GPU timing overheads. Removing the synchronization between iterations may in theory be fine if we really think we need it (which I'm not sure); however, I've been doing some testing - this may lead to occasional segfaults under load in throughput mode (with 16-32 threads or so). I haven't been able to identify the root of the problem so far.

Reply:
Moving these outside the loop was meant as a quick check on how much overhead is added. Currently we measure `latency = GPU time + synch overhead`. Question is whether sync overhead shall be part of the equation or not. I would prefer to keep `gpu_timer.lap()` in the loop, and possibly make it optional.

Reply:
Removed this change.